domain/keymanager/state/state.go

```diff
@@ -33,196 +35,415 @@ func NewState(factory database.TxnRunnerFactory) *State {
-				continue
-			} else if err != nil {
-				return err
+			keyId, err := s.ensureUserPublicKey(ctx, row, tx)
+			if err != nil {
+				return errors.Errorf(
+					"ensuring user %q public key %d on model %q: %w",
+					userUUID, i, modelUUID, err,
+				)
 			}
+			keyIds = append(keyIds, keyId)
 		}
+
+		for i, keyId := range keyIds {
+			row := modelAuthorizedKey{
+				UserPublicSSHKeyId: keyId,
+				ModelUUID:          modelUUID.String(),
+			}
+
+			err := tx.Query(ctx, insertModelAuthorisedKeyStmt, row).Run()
+			if err != nil && !jujudb.IsErrConstraintPrimaryKey(err) {
+				return errors.Errorf(
+					"ensuring key %d for user %q on model %q: %w",
+					i, userUUID, modelUUID, err,
+				)
+			}
+		}
+
 		return nil
 	})
 
-	if err != nil {
-		return fmt.Errorf(
-			"cannot add %d keys for user %q: %w",
-			len(publicKeys), userId, err,
-		)
-	}
 	return err
 }
 
-// GetPublicKeysForUser is responsible for returning all of the public keys for
-// the user id in this model. If the user does not exist no error is returned.
+// GetPublicKeysForUser is responsible for returning all of the public
+// keys for the user uuid on a model. If the user does not exist no error is
+// returned.
+// The following errors can be expected:
+// - [accesserrors.UserNotFound] if the user does not exist.
+// - [modelerrors.NotFound] if the model does not exist.
 func (s *State) GetPublicKeysForUser(
 	ctx context.Context,
-	id user.UUID,
+	modelUUID model.UUID,
+	userUUID user.UUID,
 ) ([]coressh.PublicKey, error) {
 	db, err := s.DB()
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, err
 	}
 
-	stmt, err := s.Prepare(`
-SELECT (public_key, fingerprint) AS (&publicKey.*)
-FROM user_public_ssh_key
-WHERE user_id = $userId.user_id
-`, userId{}, publicKey{})
+	modelUUIDVal := modelUUIDValue{UUID: modelUUID.String()}
+	userUUIDVal := userUUIDValue{UUID: userUUID.String()}
 
+	stmt, err := s.Prepare(`
+SELECT (upsk.public_key, upsk.fingerprint) AS (&publicKey.*)
+FROM user_public_ssh_key AS upsk
+INNER JOIN model_authorized_keys AS m ON upsk.user_public_ssh_key_id = m.user_public_ssh_key_id
+WHERE user_uuid = $userUUIDValue.user_uuid
+AND model_uuid = $modelUUIDValue.model_uuid
+`, userUUIDVal, publicKey{}, modelUUIDVal)
 	if err != nil {
-		return nil, fmt.Errorf(
-			"preparing select statement for getting public keys of user %q: %w",
-			id, err,
+		return nil, errors.Errorf(
+			"preparing select statement for getting public keys of user %q on model %q: %w",
+			userUUID, modelUUID, err,
 		)
 	}
 
-	userId := userId{UserId: id.String()}
 	publicKeys := []publicKey{}
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		err := tx.Query(ctx, stmt, userId).GetAll(&publicKeys)
-		if errors.Is(err, sqlair.ErrNoRows) {
-			return nil
-		} else if err != nil {
-			return err
+		err := s.checkUserExists(ctx, userUUID, tx)
+		if err != nil {
+			return errors.Errorf(
+				"getting public keys for user %q on model %q: %w",
```

Comment:
"checking if the user %q exists on model %q" might be better.

Reply:
That context will be in the %w

Reply:
Done & Done